src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java

```diff
@@ -0,0 +1,268 @@
+
+        //Get unmapped reads (note these always come in pairs)
+        //TODO: retain read pairs by alignment score instead of flags
+        final JavaRDD<GATKRead> readsNonHost = readsAligned.filter(read -> read.isUnmapped() && read.mateIsUnmapped());
+        logger.info("Reads remaining after BWA filtering: " + readsFilteredPaired.count());
+
+        //TODO: repeat BWA with seed size 11
+
+        //Write output
+        header.setSortOrder(SAMFileHeader.SortOrder.queryname);
+        try {
+            ReadsSparkSink.writeReads(ctx, OUTPUT_PATH, null, readsNonHost, header, shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE);
+        } catch (final IOException e) {
+            throw new GATKException("Unable to write bam",e);
+        }
+    }
+
+    private JavaRDD<GATKRead> doHostBWA(final JavaSparkContext ctx, final SAMFileHeader readsHeader, final JavaRDD<GATKRead> reads) {
+
+        final BwaSparkEngine engine = new BwaSparkEngine(bwaArgs.numThreads, bwaArgs.fixedChunkSize, HOST_REF_PATH);
+        final GCSOptions gcsOptions = getAuthenticatedGCSOptions(); // null if we have no api key
+        final ReferenceMultiSource hostReference = new ReferenceMultiSource(gcsOptions, HOST_REF_PATH, getReferenceWindowFunction());
+        final SAMSequenceDictionary hostRefDict = hostReference.getReferenceSequenceDictionary(header.getSequenceDictionary());
+        readsHeader.setSequenceDictionary(hostRefDict);
+        return engine.alignWithBWA(ctx, reads, readsHeader);
+    }
+
+    @SuppressWarnings("unchecked")
+    private JavaRDD<GATKRead> doKmerFiltering(final JavaSparkContext ctx, final JavaRDD<GATKRead> reads) {
+
+        final PipelineOptions options = getAuthenticatedGCSOptions();
+        Input input = new Input(BucketUtils.openFile(KMER_LIB_PATH, options));
+        Kryo kryo=new Kryo();
+        kryo.setReferences(false);
+
+        Set<SVKmer> kmerLibSet = (HopscotchSet<SVKmer>)kryo.readClassAndObject(input);
+
+        return reads.filter(new ContainsKmerReadFilterSpark(ctx.broadcast(kmerLibSet),KMER_SIZE));
+    }
+
+    private static JavaRDD<GATKRead> retainPairs(final JavaSparkContext ctx, JavaRDD<GATKRead> reads) {
+        JavaRDD<GATKRead> pairedReads = reads.groupBy(read -> read.getName())
+                                            .values()
+                                            .filter(p -> {
+                                                final Iterator<GATKRead> itr = p.iterator();
+                                                itr.next();
+                                                return itr.hasNext();})
+                                            .flatMap(pairedRead -> pairedRead);
+        return pairedReads;
+    }
+
+    /**
+     * This version of retainPairs() groups read name Strings instead of the GATKReads themselves for faster shuffling.
+     */
+    private static JavaRDD<GATKRead> retainPairs_ShuffleByName(final JavaSparkContext ctx, JavaRDD<GATKRead> reads) {
```

Comment:
Yes I'd like to keep it for future performance testing on a distributed architecture.